### PR TITLE
perf: speed up parsing by up to 60%

### PR DESCRIPTION
**Pull request by lukecotter**
Up to 60% faster parsing. The biggest improvement is for larger files.
The changes are around resolving the line indexes from character indexes and so most changes are in `resolvelineindexes`.
I implemented a binary search to find the line numbers for the given `startIndex` and `endIndex`

## Test numbers

**95 line Apex class**
before: 1087.75 ms
after: 1068.5 ms

**528 line Apex class**
before: 190 ms
after: 156.5 ms

**11_230 line Apex class**
before: 5710.25 ms
after: 3434.75 ms

## Images
Produced using `node --inspect-brk` and chrome profiling.
The Apex Class used was 11_230 lines long

**Before**
prettier-apex-perf-before

**After**
prettier-apex-perf-after

- [X] I’ve added tests to confirm my change works.

This was a perf improvement but I did testing and have attached before and after images of the flame chart as well as before and after numbers of time taken.

- [X] (If changing formatting output/API or CLI) I’ve added my changes to the `CHANGELOG.md` file in the `Unreleased` section.
- [X] I’ve read the [contributing guidelines](https://github.com/dangmai/prettier-plugin-apex/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by lukecotter** [comment]
Further improvements have gotten the 11_230 line number between 3200 and 3300 ms

**Comment by dangmai** [comment]
Somehow I missed the notification for this PR! Thank you very much for taking a look into this 💯 , this part of the code hasn't been touched in a while and I know there's plenty of opportunities to make it faster.
There are currently some linting issues, do you mind fixing them?

**Comment by lukecotter** [comment]
Linting issues all fixed 👍🏻 

I had to switch from for..of back to forEach, due to the rule _"iterators/generators require regenerator-runtime, which is too heavyweight for this guide to allow them. Separately, loops should be avoided in favor of array iterations."_

In my testing for..of was faster but it was not a huge difference so happy to keep forEach if you would prefer.

File: CHANGELOG.md
Comment (dangmai) [inline CHANGELOG.md:4]:
```suggestion
- Improving parsing performance - thanks to @lukecotter for their contribution!
```

**Review by dangmai — APPROVED** [review]
small suggestions on the CHANGELOG, but the parser change LGTM! I'll merge it tomorrow once you've made the doc change. Thank you so much! 🔥

**Comment by lukecotter** [comment]
Thanks for the merge. Looking forward to faster formatting 😀

**Comment by dangmai** [comment]
> Thanks for the merge. Looking forward to faster formatting 😀

There's some exciting development in store for you then 😉  I figured out a theoretical way to make the parsing a few orders of magnitude faster. Not ready to share it publicly yet, but it's coming.

**Comment by dangmai** [comment]
I've released `v2.1.5` with your changes in it. Thanks again! 🥇

**Comment by lukecotter** [comment]
> > Thanks for the merge. Looking forward to faster formatting 😀
> 
> 
> 
> There's some exciting development in store for you then 😉  I figured out a theoretical way to make the parsing a few orders of magnitude faster. Not ready to share it publicly yet, but it's coming.

Exciting, I'll keep an eye out!
Is it still worth me experimenting with a new parser?

**Comment by dangmai** [comment]
> > > Thanks for the merge. Looking forward to faster formatting 😀
> > 
> > 
> > There's some exciting development in store for you then 😉  I figured out a theoretical way to make the parsing a few orders of magnitude faster. Not ready to share it publicly yet, but it's coming.
> 
> Exciting, I'll keep an eye out! Is it still worth me experimenting with a new parser?

I'd hold off on that for a bit, unless you just want to poke around and see how Prettier Apex works 😄

**Pull request merged**
